Support IBC

KamiD opened the pull request:
<!-- < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < ☺
v                               ✰  Thanks for creating a PR! ✰    
v    Before smashing the submit button please review the checkboxes.
v    If a checkbox is n/a - please still include it but + a little note why
☺ > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > >  -->

- Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/okex/okexchain/blob/develop/CONTRIBUTING.md#pr-targeting))

- [ ] Linked to github-issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Wrote tests
- [ ] Updated relevant documentation (`docs/`)
- [ ] Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [ ] Reviewed `Files changed` in the github PR explorer

______

For Admin Use:
- Added appropriate labels to PR (ex. wip, ready-for-review, docs)
- Reviewers Assigned
- Squashed all commits, uses message "Merge PR #XYZ: [title]" ([coding standards](https://github.com/okex/coding/blob/master/README.md#merging-a-pr))

Commit:
add ibc-go module v2.0.3

Commit:
add proto.pb cosmos dependence file

Commit:
ibc-dev-merged

louisliu2048 commented on libs/iavl/mutable_tree_oec.go:
```diff
@@ -53,8 +53,10 @@ func (tree *MutableTree) SaveVersionAsync(version int64, useDeltas bool) ([]byte
 		if useDeltas {
 			tree.updateBranchWithDelta(tree.root)
-		} else {
+		} else if produceDelta {
```
global var?

Ywmet replied:
it is from commit bfcbb638e741964baab783caaca28d0214692db7

ItsFunny committed:
rename ports
interceptor
codes optimize
remove notes dev/ibc dir protos codes duplicate
resp
ibcReDnm
rename getibcheight
add proto, cdcWrapper

louisliu2048 commented on libs/cosmos-sdk/store/iavl/iavl_store.go:
```diff
@@ -566,3 +579,11 @@ func (st *Store) Import(version int64) (*iavl.Importer, error) {
 	return tree.Import(version)
 }
+
+func (st *Store) UpgradeVersion(version int64) {
```
maybe `SetUpgrateVersion` func name will be better

louisliu2048 commented on libs/cosmos-sdk/store/rootmulti/rootmulti_store.go:
```diff
@@ -64,11 +70,16 @@ type Store struct {
 var (
 	_ types.CommitMultiStore = (*Store)(nil)
 	_ types.Queryable        = (*Store)(nil)
+	_ types.CommitMultiStore = (*Store)(nil)
```
duplicate line

Ywmet committed:
Merge pull request #1789 from okex/dev-ibc-base0331

rename ports

zhongqiuwood committed:
Merge branch 'dev' into dev-ibc-base

Ywmet committed:
address ibc review comments (#1792)

* rename ports

* interceptor

* codes optimize

* remove notes

* remove protos
remove dev/ibc dir

resp

* remove codes

* add proto

* ibcReDnm

* remove file

* rename getibcheight

* remove codes

* add proto, cdcWrapper

* remove codes

* remove duplicate

* fix

* add tx decoder ut

* add tx decoder ut

* optimize evm hook

* err nil order

* SetUpgradeVersion

* fix and optimize

* remove codes

* remove notes

* task

* Update base.go

* add txDecode ut refer to evm/types/msg_test.go

* revert hard code

* delete 'OKCString' function

* Dev ibc base0401 fix repair state (#1799)

* fix repair-state

* add global value for repair ibc upgrade state when height lower than Venesis1Height

* lrp

* repair state ok

* tidy milestone

Co-authored-by: itsfunny <18757883747@163.com>
Co-authored-by: Ywmet <xiaoliang.zhang@okg.com>
Co-authored-by: Zhong Qiu <36867992+zhongqiuwood@users.noreply.github.com>

* optimize codec (#1796)

* optimize codec

* fix restful

* optmize

* remove

* task seal

* codec

* msg type name
rename view msg wrapper

Co-authored-by: KamiD <44460798+KamiD@users.noreply.github.com>

* rm upgrade Venus1 logic,no use,others ibc upgrade for fit ibc keeper interface (#1807)

Co-authored-by: Ywmet <xiaoliang.zhang@okg.com>

Co-authored-by: itsfunny <18757883747@163.com>
Co-authored-by: Ywmet <xiaoliang.zhang@okg.com>
Co-authored-by: ylsGit <bzwdyls@163.com>
Co-authored-by: Zhong Qiu <36867992+zhongqiuwood@users.noreply.github.com>
Co-authored-by: ItsFunny <39111451+ItsFunny@users.noreply.github.com>
Co-authored-by: KamiD <44460798+KamiD@users.noreply.github.com>

Ywmet committed:
fix uts (#1809)

* rm ut

* fix app

* fix tm ut

* fix app

* height 0

* venus height 0

* Makefile

* fix

* fix app ut

* upgrade task height

* fix codec ut

* lrp

* remove notes

* add codec

* fix ibc tx validate
fix TestMultistoreLoadWithUpgrade
fix TestSetLoader

* optimize height

* fix ut

* fix HigherThanVenus1 to higher or equal

Co-authored-by: Ywmet <xiaoliang.zhang@okg.com>
Co-authored-by: itsfunny <18757883747@163.com>
Co-authored-by: KamiD <44460798+KamiD@users.noreply.github.com>

KamiD committed:
revert logic of HigherThanVenus1 and UpgradeHeight

KamiD committed:
add comment
